update argument checks from padding layers

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102949
* #102597
* #102518
* #102255
* #102254
* __->__ #102253

replacement of https://github.com/pytorch/pytorch/pull/99608, breaking the old pr into smaller ones.

this one handles the common error message from both CPU and CUDA device, to simplify the code.

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

mingfeima committed:
update argument checks from padding layers

[ghstack-poisoned]

mingfeima committed:
Update on "update argument checks from padding layers"


replacement of https://github.com/pytorch/pytorch/pull/99608, breaking the old pr into smaller ones.

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "update argument checks from padding layers"


replacement of https://github.com/pytorch/pytorch/pull/99608, breaking the old pr into smaller ones.

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented:
@cpuhrsch could you please help review this stack again, i have broke #99608 into several smaller PRs. Thanks!

cpuhrsch commented on aten/src/ATen/native/ReplicationPadding.cpp:
```diff
@@ -220,6 +161,11 @@ TORCH_META_FUNC(replication_pad3d) (
   int64_t oheight = iheight + ptop + pbottom;
   int64_t owidth  = iwidth + pleft + pright;
 
+  TORCH_CHECK(owidth >= 1 || oheight >= 1 || odepth >= 1,
```
Is this constraint new or something we just failed to check so far?

mingfeima replied:
This part is a copy paste from https://github.com/pytorch/pytorch/pull/102253/files/e854341982367927e1792b37b172bba7d1f0142d#diff-1e3e579837dd1291362434c7935fbe62601e5b3511820a7e6507fe4d8fbd879bL178

the error message is identical to previous.

cpuhrsch commented on aten/src/ATen/native/ReplicationPadding.cpp:
```diff
@@ -761,9 +706,15 @@ Tensor& replication_pad3d_backward_out_cpu_template(
   int64_t owidth  = iwidth + pleft + pright;
 
-
-  shapeCheck3d(input, pleft, pright,
```
Do you need to add a call to `at::native::padding::check_valid_input` here too?

mingfeima replied:
I don't think so.

The original code does check with `shapeCheck3d` on both forward and backward.

But `shapeCheck3d` is checking the `input`, I don't think we need to re-check the `input` from a backward calling. The backward is not intended to be used directly by users, it is launched by autograd, we should have already checked the `input` in the forward. So i think the original check of input from the backward is a duplicate.

cpuhrsch replied:
@albanD - what do you think?

mingfeima replied:
fair enough, the input check is added back.

cpuhrsch commented on aten/src/ATen/native/ReflectionPad.cpp:
```diff
@@ -81,14 +77,7 @@ TORCH_META_FUNC(reflection_pad1d)(const Tensor& input, IntArrayRef padding) {
-  int64_t dim_w = 1;
-  int64_t nbatch = 1;
-
-  if (input.ndimension() == 3) {
```
Hm, how come this was possible to be removed?

mingfeima replied:
The thing is since we directly use a kernel to implement `reflection_pad1d_backward`, we don't need to get `nbatch` any more. Just need to get `dim_w` in order to check the `input_w`.

We may have two types inputs:
* batch mode: NCW
* non batch mode: CW
 
in either type, `dim_w` is the last dimension, so we can simplify the code with a single line of
`int64_t dim_w = input.dim() - 1;`

cpuhrsch replied:
Hm, my only worry is that this logic will break down if there's no dimension check. 

I know the documentation says there's only these two cases, but from what I can tell the code doesn't enforce it. 

Adding explicit checks also doesn't seem to be worth the trouble.

Can we just keep the current logic? I don't think cleaning this up is worth the potential fallout.

cpuhrsch replied:
@mingfeima - This is the comment I meant that's left to resolve.

mingfeima replied:
Oh sure, the PR is updated!

mingfeima committed:
Update on "update argument checks from padding layers"


replacement of https://github.com/pytorch/pytorch/pull/99608, breaking the old pr into smaller ones.

this one handles the common error message from both CPU and CUDA device, to simplify the code.

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented:
updated ! input check for backward is added back !

cpuhrsch reviewed (changes requested):
Only one comment left to resolve. Otherwise this looks good.

albanD reviewed (approved):
SGTM!

mingfeima committed:
Update on "update argument checks from padding layers"


replacement of https://github.com/pytorch/pytorch/pull/99608, breaking the old pr into smaller ones.

this one handles the common error message from both CPU and CUDA device, to simplify the code.

cc jgong5 XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge
